CRM/Core/SelectValues.php

```diff
@@ -1203,4 +1204,25 @@ public static function relationshipOrientation() {
     ];
   }
 
+  /**
+   * @return array
+   */
+  public static function timezone() {
+    static $tzlist = NULL;
+
+    if (is_null($tzlist)) {
+      $tzlist = [];
+      foreach (timezone_identifiers_list() as $tz) {
+        // Actual timezone keys for PHP are mapped to translated parts.
+        $tzlist[$tz] = implode('/', array_map('ts', explode('/', str_replace('_', ' ', $tz))));
```

Comment:
I'm not sure about this use of `ts` - the strings to be translated need to be literals in the code _somewhere_ for them to get pulled out for translators to work on.

Reply:
I see what you're saying, though I think we really ought to have translation for these.  I think my idea here was that if there was some hook or equivalent we could use to deliver the timezone names to the translation files we could use that... but I never found such a hook.

Reply:
Right - two thoughts:

1. Have Civi maintain its own list of these as option values. Keep them updated as climate change causes time itself to warp and the zones change.
2. Since it's coming from a built-in php function, it may already be translated if the operating system is using a non-english locale. That won't work for multilingual, but is maybe something to think about anyway since it means ts() would fail since the source always needs to be en_US. So does (1) sound crazy?

Reply:
Ok I might be wrong about (2). But (1) would help with multilingual, and as a bonus you'd get to run that cool regen script again.

Reply:
The built-in function comes with machine names only, and there's 425 of these at current count (and who knows what State boundaries will change in the hypothetical upcoming conflict between certain highly influential countries that don't seem to be able to leave each other alone)....

Since we're requiring the Intl extension anyway, another option is to using the IntlDateFormatter class, e.g.:
```php
$tzfmt = IntlDateFormatter::create(CRM_Core_Config::singletion()->userSystem->getUFLocale(), null, null, $tz, IntlDateFormatter::GREGORIAN, 'VVVV');
$tzlist[$tz] = is_object($tzfmt) ? $tzfmt->format(0) : $tz;
```
(I'm not sure about `getUFLocale()` here – is there a better way to get the current language context? Does CiviCRM setlocale() and we can use NULL here?)

Which will return strings like "Montenegro Time" or "Crna Gora" depending on locale.

The flow on effect of this though is that we're now using two formatting functions with their own formats.

Reply:
Personally I would not mind if translation is fixed later in a subsequent PR, as long as we have some sense of how it can be done. This PR is already a ton of work.

For what it's worth, here is what WordPress does. Just posting this here for comparison/brainstorm. I think that IntlDateFormatter is a better lead:  
https://developer.wordpress.org/reference/functions/wp_timezone_choice/

![tz-2021-06-30_22-21](https://user-images.githubusercontent.com/254741/124055272-9845f780-d9f1-11eb-9993-140fac3ec176.png)

> (I'm not sure about getUFLocale() here – is there a better way to get the current language context? Does CiviCRM setlocale() and we can use NULL here?)

`$tsLocale = CRM_Core_I18n::getLocale();` will return a locale such as `fr_CA` or  `en_UK`. The "UFLocale" is the language of the CMS, which can sometimes be different than CiviCRM, but I think here `getLocale` is what you're looking for.

> "Does CiviCRM setlocale() and we can use NULL here?)"

Not reliably. CiviCRM only does a setlocale() if using "CIVICRM_GETTEXT_NATIVE", which I suspect most people don't (well, I do, but it requires making sure that the server has the proper locales setup at the OS-level).

> "IntlDateFormatter::create"

I'm not familiar with this function. VVVV is odd in French, but VVV might work?

```php
$tz = "America/Toronto";
$locale = 'ru_RU';
$tzfmt = IntlDateFormatter::create($locale, null, null, $tz, IntlDateFormatter::GREGORIAN, 'VVVV');
$tzlist[$tz] = is_object($tzfmt) ? $tzfmt->format(0) : $tz;
print_r($tzlist);
// result:
    [America/Toronto] => Торонто
```

However in French it looks awkward:

```
$tz = "America/Toronto";
$locale = 'fr_CA';
$tzfmt = IntlDateFormatter::create($locale, null, null, $tz, IntlDateFormatter::GREGORIAN, 'VVVV');
$tzlist[$tz] = is_object($tzfmt) ? $tzfmt->format(0) : $tz;
print_r($tzlist);
// result:
    [America/Toronto] => heure : Toronto
// would be equivalent to "time : Toronto"
// en_UK outputs "Toronto Time"
```

Using `VVV` seems better, but I couldn't find documentation about it:

```
$tz = "America/Toronto";
$locale = 'fr_CA';
$tzfmt = IntlDateFormatter::create($locale, null, null, $tz, IntlDateFormatter::GREGORIAN, 'VVVV');
$tzlist[$tz] = is_object($tzfmt) ? $tzfmt->format(0) : $tz;
print_r($tzlist);
// result:
    [America/Toronto] => Toronto
```

I also tested `zn_CN`, which gives "多伦多" (Toronto) instead of "多伦多时间" (Toronto time).

On the system where I tested, I only had the `fr_CA` locale, not Russian or Chinese.

The last thing missing would be to translate "America" to "Amériques", for examples, for that we could probably have a hardcoded array lookup with ts() ?

```
$regions = [
  'Africa' => ts('Africa'),
  'America' => ts('America'),
  // etc
];
```

Reply:
Documentation is at https://unicode-org.github.io/icu/userguide/format_parse/datetime/#date-field-symbol-table

VVV has the disadvantage that it's not guaranteed to be (and even in my limited testing, isn't) unique, so doesn't work super well when you're *selecting* a timezone.

I'm not actually super keen on the CLDR output anyway because the country name is still only shown if it needs to be, and even when it is shown it's later in the string, when results in a dropdown box that's a superb mess of "Where's the thing I want?"

Anyway, sounds like we just remove the naive attempt at translation for now.